File: frontend/lib/features/habits/data/habit_repository_impl.dart
import '../../../../core/errors/exceptions.dart';
import '../../../../core/errors/failure.dart';
import '../../../../core/network/api_client.dart';
import 'habit_dto.dart';
import 'habit_entry_dto.dart';
import '../domain/habit.dart';
import '../domain/habit_entry.dart';
import '../domain/habit_repository.dart';


class HabitRepositoryImpl implements HabitRepository {
  final ApiClient apiClient;

  HabitRepositoryImpl(this.apiClient);

  @override
  Future<List<Habit>> getHabits() async {
    try {
      const userId = "EtFsnFHpzxYVYuiiBzGEm354FIl2";

      final response = await apiClient.get('/users/$userId/habits');
      final List data = response.data as List;

      final now = DateTime.now();
      final yearStart = DateTime(now.year, 1, 1);
      final yearEnd = DateTime(now.year, 12, 31);

      final habits = <Habit>[];
      for (final json in data) {
        final habit = HabitDto.fromJson(json as Map<String, dynamic>).toDomain();
        final entries = await getHabitEntries(
          habit.id,
          startDate: yearStart,
          endDate: yearEnd,
        );
        habits.add(Habit(
          id: habit.id,
          userId: habit.userId,
          title: habit.title,
          frequency: habit.frequency,
          startDate: habit.startDate,
          isArchived: habit.isArchived,
          entries: entries,
        ));
      }
      return habits;
    } on ServerException catch (e) {
      throw ServerFailure(e.message);
    } on NetworkException catch (e) {
      throw NetworkFailure(e.message);
    } catch (e) {
      throw UnknownFailure(e.toString());
    }
  }

  @override
  Future<List<HabitEntry>> getHabitEntries(String habitId, {DateTime? startDate, DateTime? endDate}) async {
    try {
      var path = '/habits/$habitId/entries';
      final query = <String>[];
      if (startDate != null) query.add('startDate=${startDate.toIso8601String().split('T').first}');
      if (endDate != null) query.add('endDate=${endDate.toIso8601String().split('T').first}');
      if (query.isNotEmpty) path += '?${query.join('&')}';

      final response = await apiClient.get(path);
      final List data = response.data as List;
      return data
          .map((e) => HabitEntryDto.fromJson(e as Map<String, dynamic>).toDomain())
          .toList();
    } on ServerException catch (e) {
      throw ServerFailure(e.message);
    } on NetworkException catch (e) {
      throw NetworkFailure(e.message);
    } catch (e) {
      throw UnknownFailure(e.toString());
    }
  }

  @override
  Future<Habit> createHabit(Habit habit) async {
    try {
      // Backend HabitCreate expects only: title, frequency, startDate, userId
      final createPayload = {
        'title': habit.title,
        'frequency': habit.frequency,
        'startDate': habit.startDate.toIso8601String().split('T').first,
        'userId': habit.userId,
      };

      final response = await apiClient.post('/habits', data: createPayload);
      return HabitDto.fromJson(response.data as Map<String, dynamic>).toDomain();
    } on ServerException catch (e) {
      throw ServerFailure(e.message);
    } on NetworkException catch (e) {
      throw NetworkFailure(e.message);
    } catch (e) {
      throw UnknownFailure(e.toString());
    }
  }

  @override
  Future<HabitEntry> addHabitEntry(
      String habitId, HabitEntry entry) async {
    try {
      final payload = {
        'habitId': habitId,
        'date': entry.date,
        'completed': entry.completed,
      };
      final response = await apiClient.post(
        '/habit-entries',
        data: payload,
      );

      return HabitEntryDto.fromJson(response.data as Map<String, dynamic>).toDomain();
    } on ServerException catch (e) {
      throw ServerFailure(e.message);
    } on NetworkException catch (e) {
      throw NetworkFailure(e.message);
    } catch (e) {
      throw UnknownFailure(e.toString());
    }
  }

  @override
  Future<void> deleteHabit(String habitId) async {
    try {
      await apiClient.delete('/habits/$habitId');
    } on ServerException catch (e) {
      throw ServerFailure(e.message);
    } on NetworkException catch (e) {
      throw NetworkFailure(e.message);
    } catch (e) {
      throw UnknownFailure(e.toString());
    }
  }

  @override
  Future<void> toggleHabitCheck(
      String habitId,
      bool completed,
      ) async {
    try {
      final now = DateTime.now();
      final dateOnly = "${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}";
      await apiClient.post(
        '/habits/check',
        data: {
          "habitId": habitId,
          "date": dateOnly,
          "completed": completed,
        },
      );
    } on ServerException catch (e) {
      throw ServerFailure(e.message);
    } on NetworkException catch (e) {
      throw NetworkFailure(e.message);
    } catch (e) {
      throw UnknownFailure(e.toString());
    }
  }
}

File: app/src/main/java/com/example/cozytrack/data/repository/HabitRepositoryImpl.kt
package com.example.cozytrack.data.repository

import com.example.cozytrack.core.network.ApiResult
import com.example.cozytrack.core.session.SessionManager
import com.example.cozytrack.data.remote.HabitTrackerApi
import com.example.cozytrack.data.remote.dto.ClockResponseDto
import com.example.cozytrack.data.remote.dto.CreateHabitRequestDto
import com.example.cozytrack.data.remote.dto.HabitCheckRequestDto
import com.example.cozytrack.data.remote.dto.HabitDto
import com.example.cozytrack.data.remote.dto.HabitEntryDto
import com.example.cozytrack.data.remote.dto.HabitProgressDto
import com.example.cozytrack.data.remote.dto.HabitUpdateRequestDto
import com.example.cozytrack.domain.model.Frequency
import com.example.cozytrack.domain.model.Habit
import com.example.cozytrack.domain.model.HabitEntry
import com.example.cozytrack.domain.model.HabitProgress
import com.example.cozytrack.domain.model.ProgressSnapshot
import com.example.cozytrack.domain.model.ServerClock
import com.example.cozytrack.domain.repository.HabitRepository
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import javax.inject.Inject
import kotlinx.coroutines.flow.first
import retrofit2.HttpException

class HabitRepositoryImpl @Inject constructor(
    private val api: HabitTrackerApi,
    private val sessionManager: SessionManager
) : HabitRepository {
    override suspend fun getServerClock(): ApiResult<ServerClock> {
        return safeCall { api.getClock().toDomain() }
    }

    override suspend fun getHabits(includeArchived: Boolean): ApiResult<List<Habit>> {
        return safeCall {
            val userId = sessionManager.userId.first()
                ?: error("You must be logged in to load habits")

            api.getHabits(userId = userId, includeArchived = includeArchived)
                .map { it.toDomain() }
        }
    }

    override suspend fun getHabit(habitId: String): ApiResult<Habit> {
        return safeCall {
            api.getHabit(habitId).toDomain()
        }
    }

    override suspend fun createHabit(
        title: String,
        frequency: Frequency,
        startDate: String
    ): ApiResult<String> {
        return safeCall {
            val userId = sessionManager.userId.first()
                ?: error("You must be logged in to create habits")

            api.createHabit(
                CreateHabitRequestDto(
                    title = title,
                    frequency = frequency.apiValue,
                    startDate = startDate,
                    userId = userId
                )
            ).string()
        }
    }

    override suspend fun updateHabit(
        habitId: String,
        title: String?,
        frequency: Frequency?,
        startDate: String?,
        isArchived: Boolean?
    ): ApiResult<Unit> {
        return safeCall {
            api.updateHabit(
                habitId = habitId,
                request = HabitUpdateRequestDto(
                    title = title,
                    frequency = frequency?.apiValue,
                    startDate = startDate,
                    isArchived = isArchived
                )
            ).close()
        }
    }

    override suspend fun deleteHabit(habitId: String): ApiResult<Unit> {
        return safeCall {
            val response = api.deleteHabit(habitId)
            if (!response.isSuccessful) {
                val errorBody = response.errorBody()?.string().orEmpty()
                error(
                    if (errorBody.isBlank()) {
                        "HTTP ${response.code()}"
                    } else {
                        "HTTP ${response.code()}: $errorBody"
                    }
                )
            }
        }
    }

    override suspend fun checkHabit(habitId: String, completed: Boolean): ApiResult<Unit> {
        return safeCall {
            api.checkHabit(
                HabitCheckRequestDto(
                    habitId = habitId,
                    completed = completed
                )
            ).close()
        }
    }

    override suspend fun getEntriesForDay(checkDate: String): ApiResult<List<HabitEntry>> {
        return safeCall {
            val userId = sessionManager.userId.first()
                ?: error("You must be logged in to load entries")

            api.getEntriesForDay(userId = userId, checkDate = checkDate)
                .map { it.toDomain() }
        }
    }

    override suspend fun getHabitEntries(
        habitId: String,
        startDate: String?,
        endDate: String?
    ): ApiResult<List<HabitEntry>> {
        return safeCall {
            api.getHabitEntries(
                habitId = habitId,
                startDate = startDate,
                endDate = endDate
            ).map { it.toDomain() }
        }
    }

    override suspend fun getHabitProgress(
        habitId: String,
        startDate: String?,
        endDate: String?
    ): ApiResult<HabitProgress> {
        return safeCall {
            api.getHabitProgress(
                habitId = habitId,
                startDate = startDate,
                endDate = endDate
            ).toDomain()
        }
    }

    private suspend fun <T> safeCall(call: suspend () -> T): ApiResult<T> {
        return try {
            ApiResult.Success(call())
        } catch (error: Exception) {
            ApiResult.Error(error.toApiMessage())
        }
    }
}

private fun Exception.toApiMessage(): String {
    if (this is UnknownHostException) {
        return "Cannot reach the CozyTrack server. Check that your emulator/device has internet access and DNS is working."
    }

    if (this is SocketTimeoutException) {
        return "The CozyTrack server took too long to respond. Please try again."
    }

    if (this is HttpException) {
        val errorBody = response()?.errorBody()?.string().orEmpty()
        return if (errorBody.isBlank()) {
            "HTTP ${code()}"
        } else {
            "HTTP ${code()}: $errorBody"
        }
    }

    return message ?: "Something went wrong"
}

private fun ClockResponseDto.toDomain(): ServerClock {
    return ServerClock(
        utcCalendarDate = utcCalendarDate,
        utcDateTime = utcDateTime,
        timezone = timezone
    )
}

private fun HabitDto.toDomain(): Habit {
    return Habit(
        id = id,
        title = title,
        frequency = Frequency.fromApi(frequency),
        startDate = startDate,
        isArchived = isArchived
    )
}

private fun HabitEntryDto.toDomain(): HabitEntry {
    return HabitEntry(
        id = id,
        habitId = habitId,
        date = date.ifBlank { checkDate ?: calendarDate.orEmpty() },
        completed = completed,
        completedAt = completedAt
    )
}

private fun HabitProgressDto.toDomain(): HabitProgress {
    val frequency = Frequency.fromApi(frequency)
    val snapshots = when (frequency) {
        Frequency.Daily -> last14Days
        Frequency.Weekly -> last14Weeks
    }

    return HabitProgress(
        habitId = habitId,
        title = title,
        frequency = frequency,
        rangeStart = rangeStart,
        rangeEnd = rangeEnd,
        scheduledOpportunities = scheduledOpportunities,
        doneCount = doneCount,
        missedCount = missedCount,
        skippedCount = skippedCount,
        completionRate = completionRate,
        currentStreak = currentStreak,
        bestStreak = bestStreak,
        snapshots = snapshots.map {
            ProgressSnapshot(
                calendarDate = it.calendarDate,
                status = it.status
            )
        }
    )
}
